javalib-ex/src/main/scala/java/io/DataInputStream.scala

```diff
@@ -69,24 +54,27 @@ class DataInputStream(in: InputStream) extends FilterInputStream(in)
   }
 
   def readChar(): Char = {
-    if (hasArrayBuffer)
+    if (hasArrayBuffer) {
       bufDataView.getUint16(consumePos(2)).toChar
-    else
-      view(2).getUint16(0).toChar
+    } else {
+      ((readByte() << 8) | (readByte() & 0xff)).toChar
+    }
   }
 
   def readDouble(): Double = {
-    if (hasArrayBuffer)
+    if (hasArrayBuffer) {
       bufDataView.getFloat64(consumePos(8))
-    else
-      view(8).getFloat64(0)
+    } else {
+      java.lang.Double.longBitsToDouble(readLong())
```

Comment:
I'm not sure it matters. It will complicate the code quite a bit, and the two boolean checks are not a lot to pay IMO...